### ISSUE TITLE
type_contents_order doesn't follow the `order` setting for deinit

### ISSUE THREAD

**Issue by bsarrazin** [realm/SwiftLint#3165]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

```yaml
type_contents_order:
  order:
    - SwiftLintFramework.TypeContent.associatedType
    - SwiftLintFramework.TypeContent.typeAlias
    - SwiftLintFramework.TypeContent.subtype
    - SwiftLintFramework.TypeContent.typeProperty
    - SwiftLintFramework.TypeContent.typeMethod
    - SwiftLintFramework.TypeContent.ibInspectable
    - SwiftLintFramework.TypeContent.ibOutlet
    - SwiftLintFramework.TypeContent.case
    - SwiftLintFramework.TypeContent.instanceProperty
    - SwiftLintFramework.TypeContent.initializer        // inits followed by
    - SwiftLintFramework.TypeContent.deinitializer      // deinits creates false warnings
    - SwiftLintFramework.TypeContent.viewLifeCycleMethod
    - SwiftLintFramework.TypeContent.subscript
    - SwiftLintFramework.TypeContent.ibAction
    - SwiftLintFramework.TypeContent.otherMethod
```

False warnings:
![image](https://user-images.githubusercontent.com/1077441/78300172-107af000-7505-11ea-8b29-25f65bc7435c.png)

![image](https://user-images.githubusercontent.com/1077441/78300344-55068b80-7505-11ea-9b90-76cbbeadbc89.png)

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.39.1
* Installation method used (Homebrew, CocoaPods, building from source, etc)? CocoaPods
* Paste your configuration file:

```yml
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Configuration
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
disabled_rules:
  - force_cast
  - force_try
  - line_length
  - nesting
  - operator_whitespace

opt_in_rules:

# Willfully Opting In
  - anyobject_protocol
  - array_init
  - closure_end_indentation
  - closure_spacing
  - collection_alignment
  - contains_over_filter_count
  - contains_over_filter_is_empty
  - contains_over_first_not_nil
  - contains_over_range_nil_comparison
  - convenience_type
  - discouraged_object_literal
  - discouraged_optional_boolean
  - discouraged_optional_collection
  - empty_collection_literal
  - empty_count
  - empty_string
  - empty_xctest_method
  - enum_case_associated_values_count
  - expiring_todo
  - explicit_init
  - explicit_self
  - fallthrough
  - fatal_error_message
  - file_header
  - file_name
  - file_name_no_space
  - first_where
  - flatmap_over_map_reduce
  - identical_operands
  - indentation_width
  - joined_default_parameter
  - last_where
  - legacy_multiple
  - legacy_random
  - let_var_whitespace
  - literal_expression_end_indentation
  - lower_acl_than_parent
  - modifier_order
  - multiline_arguments
  - multiline_arguments_brackets
  - multiline_function_chains
  - multiline_literal_brackets
  - multiline_parameters
  - multiline_parameters_brackets
  - nimble_operator
  - nslocalizedstring_key
  - nslocalizedstring_require_bundle
  - number_separator
  - operator_usage_whitespace
  - overridden_super_call
  - override_in_extension
  - prefer_self_type_over_type_of_self
  - private_action
  - private_outlet
  - prohibited_super_call
  - quick_discouraged_call
  - quick_discouraged_focused_test
  - quick_discouraged_pending_test
  - raw_value_for_camel_cased_codable_enum
  - reduce_into
  - redundant_nil_coalescing
  - redundant_type_annotation
  - required_enum_case
  - single_test_class
  - sorted_first_last
  - sorted_imports
  - static_operator
  - strong_iboutlet
  - toggle_bool
  - type_contents_order
  - unavailable_function
  - unneeded_parentheses_in_closure_argument
  - unused_declaration
  - unused_import
  - vertical_parameter_alignment_on_call
  - vertical_whitespace_between_cases
  - vertical_whitespace_closing_braces
  - yoda_condition

# Willfully Opting Out
  # - attributes
  # - closure_body_length
  # - conditional_returns_on_newline
  # - explicit_acl
  # - explicit_enum_raw_value
  # - explicit_top_level_acl
  # - explicit_type_interface
  # - extension_access_modifier
  # - force_unwrapping
  # - function_default_parameter_at_end
  # - implicit_return
  # - implicitly_unwrapped_optional
  # - missing_docs
  # - no_extension_access_modifier
  # - no_grouping_extension
  # - object_literal
  # - optional_enum_case_matching
  # - pattern_matching_keywords
  # - prefixed_toplevel_constant
  # - prohibited_interface_builder
  # - required_deinit
  # - switch_case_on_newline
  # - unowned_variable_capture
  # - untyped_error_in_catch
  # - vertical_whitespace_opening_braces
  # - xct_specific_matcher

included:
  - Leka
  - LekaTests
  - LekaUITests
  - DomainKit
  - TestKit

excluded:
  - Pods
  - Leka/Source/Generated/*.swift

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Customization
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cyclomatic_complexity:
  ignores_case_statements: true

identifier_name:
  excluded: [
    'by', 'fb', 'from', 'hh', 'h1', 'h2', 'i', 'id', 'in',
    'mm', 'nb', 'no', 'of', 'ok', 'on', 'op', 'r', 'ss',
    'to', 'tz', 'us', 'vc', 'with', 'x', 'y'
  ]

indentation_width:
 include_comments: false

number_separator:
  minimum_length: 5

private_outlet:
  allow_private_set: true

trailing_whitespace:
  ignores_empty_lines: false

type_contents_order:
  order:
    - SwiftLintFramework.TypeContent.associatedType
    - SwiftLintFramework.TypeContent.typeAlias
    - SwiftLintFramework.TypeContent.subtype
    - SwiftLintFramework.TypeContent.typeProperty
    - SwiftLintFramework.TypeContent.typeMethod
    - SwiftLintFramework.TypeContent.ibInspectable
    - SwiftLintFramework.TypeContent.ibOutlet
    - SwiftLintFramework.TypeContent.case
    - SwiftLintFramework.TypeContent.instanceProperty
    - SwiftLintFramework.TypeContent.initializer
    - SwiftLintFramework.TypeContent.deinitializer
    - SwiftLintFramework.TypeContent.viewLifeCycleMethod
    - SwiftLintFramework.TypeContent.subscript
    - SwiftLintFramework.TypeContent.ibAction
    - SwiftLintFramework.TypeContent.otherMethod

type_name:
  excluded: ['T']
  max_length:
    warning: 50
    error: 50

```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? no
  If so, paste their relative paths and respective contents.
* Which Xcode version are you using (check `xcodebuild -version`)? 11.3.1
* Do you have a sample that shows the issue? see above

**Comment by VladNaimark** [comment]
Have same issue. No nested configs. Sample to reproduce the issue:

class Test {
    init() {}
    
    deinit {}
    
    func doNothing() {}
}

**Comment by fredpi** [comment]
I just experienced the same issue and investigated it. Here are my results:

- @bsarrazin The reason why you experienced the issue is probably because you are using the `SwiftLintFramework.TypeContent.` prefix when specifying the order of the types in your `.swiftlint.yml`. I checked the SwiftLint source code and it expects e. g. `deinitializer` instead of `SwiftLintFramework.TypeContent.deinitializer`.
- @VladNaimark What you are experiencing is probably expected behavior. You don't seem to have your own configuration that specifies the order of types and by default, the deinitializer is expected to come at the very end. **Notably, this behavior has not always been there** and has only been changed with [this PR](https://github.com/realm/SwiftLint/pull/3042/files), that did not only add the differentiation between `initializer` and `deinitializer`, but also changed the default configuration which expects the deinitializer at the end since then.

I'll close this issue, but feel free to open a new one if you believe that the default behavior should be different. Maybe @Jeehut (original creator of the `type_contents_order` rule) also has some thoughts what should be the default:
1. deinitializer at the very end
_or_
2. deinitializer after the initializers

**Comment by Jeehut** [comment]
@fredpi As I had initially not separated `deinitializer` as its own option, I think it should be implicitly clear that in my opinion the `initializer` and `deinitializer` should stay together for the same reason why I think the allocation of resources and their deallocation via `defer` should be together, too: It's harder to forget updating the "ending" part of anything "started" in an initializer if the code is close together. But having said that, I rarely implement any `deinitializer` at all, therefore **my opinion on this should not be taken as the main guide here**, maybe there might be reasons to put them to the end that I can't think of right now. I do have one possible reason though, to explain this I have to get a little more detailed:

My rationale for the order of things within a type or file follows the first fundamental principle of Swifts [API Design Guidelines](https://swift.org/documentation/api-design-guidelines/):
> Clarity at the point of use is your most important goal.

Applying this principle to the order of things for me means, that we should optimize the order for developers who might open the file to understand how to **use** it. For users of a type, I think that in general the properties of a type are most interesting (What data does it hold? / What is this?), followed by the initializer (How can I create one?) and the methods (What can I do with it?). Internal types I only prioritize as they are key to understand any of their usage in the subsequent sections, so they come even before properties. Within each part, for the same reason, the visibility should define the order (which I think we don't have a rule for yet) as a user of the type cares less about the implementation detail (the `private` stuff) than about the actual API (the `public` / `internal` stuff).

Regarding the `deinitializer`, because it typically is something a user of a type doesn't call explicitly – and therefore doesn't **use** explicitly – (the compiler decides when to call it, most of the time without the developer actually thinking about it), I would consider it a very "implementation detail" kind of thing by design. Therefore I think it can be placed very much at the bottom of a type, thus I can understand why one could consider the "`deinitializer` last" approach as a good default.

**Comment by revolter** [comment]
> The reason why you experienced the issue is probably because you are using the `SwiftLintFramework.TypeContent.` prefix when specifying the order of the types in your `.swiftlint.yml`. I checked the SwiftLint source code and it expects e.g. `deinitializer` instead of `SwiftLintFramework.TypeContent.deinitializer`.

Actually, they need to be in snake case, for example, `ib_action` instead of `ibAction`! Suprised that this isn't anywhere in the documentation.
